test/tslint-rule/anonymous-function-error.ts.lint

```diff
@@ -0,0 +1,26 @@
+const IllegalComponent = wrapper(function() {
+  useEffect(() => {});
+  ~~~~~~~~~~~~~~~~~~~ [Hook is in an anonymous function that is passed to an illegal callback. Legal callbacks identifiers that can receive anonymous functions as arguments are memo and forwardRef]
+})
+
+const LegalAnonymousComponent = function() {
+  useEffect(() => {});
+}
+
+const ForwardedComponent = React.forwardRef(function(props, ref) {
+  useEffect(() => {
+    console.log("I am legal")
+  });
+})
+
+const MemoizedComponent = React.memo((props) => {
+  const [state, setState] = React.useState(props.initValue);
+  return <span>{state}</span>
+})
+
+const Functor = function() {
+  const cb = React.useCallback(() => {
```

Comment:
After changing this line to

```diff
-  const cb = React.useCallback(() => {
+  const cb = useCallback(() => {
```

I got the following results from the test

```diff
  const Functor = function() {
    const cb = useCallback(() => {
      const r = React.useRef()
-               ~~~~~~~~~~~~~~ [Hook is in an anonymous function that is passed to an illegal callback. Legal callbacks identifiers that can receive anonymous functions as arguments are memo and forwardRef]
+               ~~~~~~~~~~~~~~ [A hook cannot be used inside of another function]
    }, [])
  }
```

This is because `React.useCallback` is not a `Identifier`, but `useCallback` is, so the condition in the `if` you added is met :worried: 

This leads me to believe that either we need to make the detection logic more robust to detect this case, or (which I still recommend), change the error message of `invalidFunctionExpression` and `invalidFunctionDeclaration` to be generic enough and still describe the problem.

An example error message could be:

> A hook cannot be used inside of a function that is not another hook or a React component. Hooks begin with "use" and React components begin with an uppercase letter

Reply:
I really believe that we shouldn't conflate these two error messages —invalid function expression and invalid anonymous function callback. Instead, what if we checked if the ancestor.parent `isComponentOrHookIdentifier`?

Reply:
Looks like it's correct now! Thanks :+1: I'll still do some more thinking about those cases, but we can definitely resolve this conversation